### PR TITLE
added coverage for edge case in pass_defender_fsm and added unit test to make sure it worked

### PR DESCRIPTION
**Pull request by TylerKerswell**
### Description
added case to set the primitive when the ball is above the robot.

### Testing Done
added unit test in pass_defender_fsm_test

### Resolved Issues
#2991 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp [hunk 1]
```diff
@@ -79,5 +79,31 @@ void PassDefenderFSM::interceptBall(const Update& event)
...
+        face_ball_orientation = Angle().zero();
+    }
+    else
+    {
+        face_ball_orientation = (ball.position() - robot_position).orientation();
     }
+
+    // backup by the length of the ball
+    Point backup_position = Point(robot_position.x() - ROBOT_MAX_RADIUS_METERS,
+                                  robot_position.y() - ROBOT_MAX_RADIUS_METERS);
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp]: I think it would be better to change the backup position relative to the velocity of the ball:

```suggestion
    Point backup_position = robot_position + ball.velocity().normalize(ROBOT_MAX_RADIUS_METERS)
```

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp [hunk 2]
```diff
@@ -49,3 +49,34 @@ TEST(PassDefenderFSMTest, test_transitions)
...
+        ::TestUtil::setBallVelocity(world, Vector(-2, 0), Timestamp::fromSeconds(123));
+    EXPECT_TRUE(world.ball().hasBallBeenKicked(Angle::half()));
+
+    // Transition to InterceptBallState
+    fsm.process_event(PassDefenderFSM::Update(
+        control_params,
+        TacticUpdate(
+            robot, world,
+            [](std::unique_ptr<TbotsProto::Primitive> x) { EXPECT_TRUE(x != nullptr); },
+            TEST_UTIL_CREATE_MOTION_CONTROL_NO_DEST)));
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp]: I would suggest rewriting this test slightly:

```suggestion
    std::unique_ptr<TbotsProto::Primitive> primitive;
    fsm.process_event(PassDefenderFSM::Update(
        control_params,
        TacticUpdate(
            robot, world,
            [&primitive](std::unique_ptr<TbotsProto::Primitive> x) { primitive = std::move(x); },
            TEST_UTIL_CREATE_MOTION_CONTROL_NO_DEST)));
    EXPECT_TRUE(primitive != nullptr);
    EXPECT_TRUE(fsm.is(boost::sml::state<PassDefenderFSM::InterceptBallState>));
```

**Review by itsarune — COMMENTED** [review]
nice work, good start!

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp (hunk 2 above)
Comment (TylerKerswell) [reply src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp]: implemented

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp (hunk 1 above)
Comment (TylerKerswell) [reply src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp]: implemented

**Review by nimazareian — COMMENTED** [review]
LGTM! Just pending the merge conflicts. Pulling from upstream master and fixing the merge conflicts should fix it (`std::unique_ptr<TbotsProto::Primitive>` was updated to `std::shared_ptr<MovePrimitive>` with some of the args moving around).

**Review by itsarune — COMMENTED** [review]
LGTM, pending merge conflicts!

**Comment by nimazareian** [comment]
@TylerKerswell Could you please fix the merge conflicts so we can get this PR in?

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp [hunk 3]
```diff
@@ -78,5 +78,30 @@ void PassDefenderFSM::interceptBall(const Update& event)
...
+    // to avoid dividing by 0
+    Angle face_ball_orientation;
+    if ((ball.position() - robot_position).length() == 0)
+    {
+        face_ball_orientation = Angle().zero();
+    }
+    else
+    {
+        face_ball_orientation = (ball.position() - robot_position).orientation();
+    }
```
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp]: I think we should make a ticket to move this logic into and handle it in the API. I suspect this may be happening in a few more places. What do you think @nimazareian ?

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp [hunk 4]
```diff
@@ -78,5 +78,30 @@ void PassDefenderFSM::interceptBall(const Update& event)
             TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT,
             TbotsProto::DribblerMode::MAX_FORCE, TbotsProto::BallCollisionType::ALLOW,
             AutoChipOrKick{AutoChipOrKickMode::OFF, 0}));
+        return;
     }
+    // something is wrong when this happens,
+    // in this special case, the robot should move away from the ball
+    // as to not violate the rule about covering more than 80% of the ball
+
+    // to avoid dividing by 0
```
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp]: You can add an `else` case to clarify the scope of this comment.

**Review by Bvasilchikov — DISMISSED** [review]
Awesome work!, grab the latest code and can't wait to see it live on the field

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp [hunk 5]
```diff
@@ -78,5 +78,30 @@ void PassDefenderFSM::interceptBall(const Update& event)
             TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT,
             TbotsProto::DribblerMode::MAX_FORCE, TbotsProto::BallCollisionType::ALLOW,
             AutoChipOrKick{AutoChipOrKickMode::OFF, 0}));
+        return;
     }
+    // something is wrong when this happens,
+    // in this special case, the robot should move away from the ball
+    // as to not violate the rule about covering more than 80% of the ball
+
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp:83]: delete empty line

**Review by itsarune — COMMENTED** [review]
looks good, pending CI and some minor nits

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp (hunk 3 above)
Comment (nimazareian) [reply src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp]: Are you referring to updating `Vector::orientation` such that it returns `Angle::zero` if its a zero-vector? That would be a good idea imo. It would be similar to `Vector::normalize`

https://github.com/UBC-Thunderbots/Software/blob/cf3b99e878db2aebf64cc083c3e74d9274dd5352/src/software/geom/vector.cpp#L52-L58

@TylerKerswell since this might be a bit off topic for your PR, could you create an issue and tag it with a TODO in your code that it should be updated?

**Comment by nimazareian** [comment]
@TylerKerswell Following up on the suggestions. Could you update this PR please?

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp [hunk 6]
```diff
@@ -78,5 +78,27 @@ void PassDefenderFSM::interceptBall(const Update& event)
...
+    if ((ball.position() - robot_position).length() == 0)
+    {
+        face_ball_orientation = event.common.robot.orientation();
+    }
+    else
+    {
+        face_ball_orientation = (ball.position() - robot_position).orientation();
+    }
+
+    // backup by the length of the ball
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm.cpp:96]:
```suggestion
    // backup by the radius of the ball
```

nit: maybe we could explain why we do this

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp [hunk 7]
```diff
@@ -41,3 +41,32 @@ TEST(PassDefenderFSMTest, test_transitions)
         control_params, TacticUpdate(robot, world, [](std::shared_ptr<Primitive>) {})));
     EXPECT_TRUE(fsm.is(boost::sml::state<PassDefenderFSM::BlockPassState>));
 }
+
+// This is created to test one single edge case in interceptBall
+TEST(PassDefenderFSMTest, test_intercept_edge_case)
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp:46]:
```suggestion
TEST(PassDefenderFSMTest, test_intercept_ball_directly_above_robot)
```

File: src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp [hunk 8]
```diff
@@ -41,3 +41,32 @@ TEST(PassDefenderFSMTest, test_transitions)
         control_params, TacticUpdate(robot, world, [](std::shared_ptr<Primitive>) {})));
     EXPECT_TRUE(fsm.is(boost::sml::state<PassDefenderFSM::BlockPassState>));
 }
+
+// This is created to test one single edge case in interceptBall
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/tactic/pass_defender/pass_defender_fsm_test.cpp:45]:
```suggestion
// This is created to test the edge case where the ball is directly above the robot in interceptBall
```

**Review by sauravbanna — CHANGES_REQUESTED** [review]

**Review by nimazareian — APPROVED** [review]

**Review by sauravbanna — APPROVED** [review]
